rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -417,6 +446,53 @@ class Publisher : public PublisherBase
     return ros_message_type_allocator_;
   }
 
+  /// Publish shared messages from intra process buffer for late joiner
+  /**
+   * This signature allows the user to give a sub_id from intra process manager.
+   * The function will publish messages currently held in buffer_ to the subscription
+   * as shared pointers.
+   *
+   * \param[in] sud_id subscription id in ipm to publish data to
+   */
+  void do_shared_intra_process_publish_for_late_joiner(const uint64_t sub_id) override
```

Comment:
these methods should be in the intra-process manager rather than in the publisher;
the publisher does not have the concept of subscription ids so it can't really use them

Reply:
I was originally planning to put the methods in intra-process manager as well, but ipm stores publishers as `PublisherBase`, and that base class has no concept of what MessageT a publisher publishes. It makes it extremely difficult to get data from a publisher buffer in intra-process manager.

Reply:
What about changing the `IntraProcessManager::add_subscription` function to be templated and take the `MessageT` as template argument?

Reply:
Mark as resolved since the design has changed.